src/api/java/com/minecolonies/api/util/InventoryUtils.java

```diff
@@ -1482,4 +1495,113 @@ public static boolean removeStacksFromProvider(final ICapabilityProvider provide
 
         return true;
     }
+
+    /**
+     * Check if a certain item is in the provider but without the provider being full.
+     * @param provider the provider to check.
+     * @param item the item.
+     * @param itemDamage its damage.
+     * @param amount stack size to be considered.
+     * @return the slot or -1.
+     */
+    public static int findFirstSlotInProviderWithNotFull(final ICapabilityProvider provider, final Item item, final int itemDamage, final int amount)
+    {
+        for (IItemHandler handler : getItemHandlersFromProvider(provider))
+        {
+            int foundSlot = findFirstSlotInItemHandlerWithNotFull(handler, (ItemStack stack) -> compareItems(stack, item, itemDamage), amount);
+            //TODO: When contract is hardened later: Replace this -1 check with a try-catch block.
+            if (foundSlot > -1)
+            {
+                return foundSlot;
+            }
+        }
+
+        return -1;
+    }
+
+    /**
+     * Check if a certain item is in the handler but without the provider being full.
+     * @param handler the handler to check.
+     * @param itemStackSelectionPredicate the selection predicate..
+     * @param amount stack size to be considered.
+     * @return the slot or -1.
+     */
+    public static int findFirstSlotInItemHandlerWithNotFull(final IItemHandler handler, @NotNull final Predicate<ItemStack> itemStackSelectionPredicate, final int amount)
+    {
+        boolean foundEmptySlot = false;
+        boolean foundItem = false;
+        int itemSlot = -1;
+        for (int slot = 0; slot < handler.getSlots(); slot++)
+        {
+            final ItemStack stack = handler.getStackInSlot(slot);
+            if(ItemStackUtils.isEmpty(stack))
+            {
+                foundEmptySlot = true;
+            }
+            else if (itemStackSelectionPredicate.test(stack))
+            {
+                if(ItemStackUtils.getSize(stack) + amount <= Constants.STACKSIZE)
+                {
+                    foundEmptySlot = true;
+                }
+                foundItem = true;
+                itemSlot = slot;
+            }
+
+            if(foundItem && foundEmptySlot)
```

Comment:
I am still not happy, with the way this method works, and what it is used for.
If I understand this correctly, it only Returns a Slot when an Empty Slot is found and when the predicate Matches, so that swaps and other operations between to inventories work.

However should it then not also return a Slot when it finds TWO empty Slots (one to fill up and one empty one)?

If not, then what the hell is this method good for.?